### ISSUE TITLE
prisma 5.6.0 `libquery_engine-darwin-arm64.dylib.node` regression

### ISSUE THREAD

**Issue by cipriancaba** [prisma/prisma#22008]
### Bug description

```
Prisma Client could not locate the Query Engine for runtime "darwin-arm64".

This is likely caused by a bundler that has not copied "libquery_engine-darwin-arm64.dylib.node" next to the resulting bundle.
Ensure that "libquery_engine-darwin-arm64.dylib.node" has been copied next to the bundle or in "node_modules/.prisma/client".

We would appreciate if you could take the time to share some information with us.
Please help us by answering a few questions: https://pris.ly/engine-not-found-bundler-investigation

The following locations have been searched:
  /Users/cipri/dev/acme/mono/apps/ops/node_modules/.prisma/client
  /Users/cipri/dev/acme/mono/apps/ops/.next/server
  /Users/cipri/dev/acme/mono/packages/db/node_modules/@prisma/client
  /Users/cipri/dev/acme/mono/apps/ops/.prisma/client
  /tmp/prisma-engines
```

### How to reproduce

[Monorepo setup](https://github.com/t3-oss/create-t3-turbo)

Prisma randomly fails with errors that it cannot find the schema or the runtime. I am using accelerate. Started happening with the latest 5.6.0

### Expected behavior

As you can see the prisma db module is used in a nextjs app called ops

The prisma script is searching everywhere except for where it actually should which is

`  /Users/cipri/dev/acme/mono/packages/db/node_modules/.prisma/client`

Not to add that the `.prisma/client` and `@prisma/client` generated folders is confusing af

### Prisma information

Not relevant

### Environment & setup

- OS: mac
- Database: postgres
- Node.js version: 18.17.1

### Prisma Version

```
5.6.0
```

**Comment by millsp** [comment]
Hey @cipriancaba, sorry to hear that you are experiencing this, you should have been pointed to a help page in the error message. Can you please visit and try https://www.prisma.io/docs/guides/other/troubleshooting-orm/help-articles/nextjs-prisma-client-monorepo, and let me know if that helped you?

**Comment by millsp** [comment]
I am curious as well about what you said. You said that you are using Accelerate. According to the error, you are not, as Accelerate is engine-less and will not search or depend on an engine. Perhaps you have passed a regular URL instead of a `prisma://...` URL?

**Comment by cipriancaba** [comment]
Hey @millsp thanks for the quick reply.

You are right, i was using accelerate but kept running into problems with transactions in server actions so reverted to direct access.

I've already done the fix, that's not the problem as this setup was working pre 5.6.0, i just upgraded from 5.5.2

As you see in the trace, which is very helpful btw, prisma is not searching in the prisma package node-modules, which should be a quick fix for this issue. It's only searching in the next app itself and the @prisma/client of the actual db package. It should search in the prisma package's node-modules/.prisma/client where the runtime gets copied

**Comment by millsp** [comment]
> As you see in the trace, which is very helpful btw, prisma is not searching in the prisma package node-modules, which should be a quick fix for this issue

I see that would help you now. The "engine search" is something that we are aware of and we know it is not perfect. In your case, the client has been bundled with Next.js. This means that in a real deployment, `db` may not actually be present in the deployment so it wouldn't help to search there. I don't want to make assumptions about your setup, maybe `db` is also deployed?

> Prisma randomly fails with errors that it cannot find the schema or the runtime.

Now cross-linking what you just told me and this. It sounds that you are maybe experiencing a bug from the webpack plugin, which sometimes "misses" to copy the engine. Could it be this? If so, we may have a solution for TS and ESM users in the coming weeks so that this plugin is not necessary at all anymore.

In the mean time, I wonder if marking `db` as external could be a solution for you?

**Comment by millsp** [comment]
If you have some time, I'd really appreciate if you could give us a reproduction that we can run easily (ready to run basically).

**Comment by cipriancaba** [comment]
Hi @millsp, thanks for your replies here, will try to setup a repro example, the only problem is that this randomly happens when I add various packages. It's really just this template with another web app, will try to create that for you

https://github.com/t3-oss/create-t3-turbo

**Comment by janpio** [comment]
Just to confirm again an avoid any confusion, this only happened when you were _not_ using Prisma Accelerate @cipriancaba, correct?
With versions before 5.6.0 you were also using Prisma directly (not via Accelerate) and you did _not_ have this problem?

**Comment by cipriancaba** [comment]
Hey @janpio and @millsp 

Here is the repro repo:

https://github.com/cipriancaba/prisma-query-lib

Just run `pnpm install` from the root

Go to the db package and run `pnpm db:generate`

Funny enough the issue doesn't seem to reproduce here because prisma client gets generated in the root node_modules as opposed to local node_modules. On my private repo it behaves differently and I am not sure why on one monorepo it's hoisted and on the other it's not. How does prisma generate choose where to generate the client by default in a monorepo?

And to answer the above:
This only happened when you were not using Prisma Accelerate, correct? -> `YES`
With versions before 5.6.0 you were also using Prisma directly (not via Accelerate) and you did not have this problem? -> `YES` but also didn't have 2 nextjs packages in the repo. So the problem started happening at some point after upgrading to 5.6.0 and adding a new next app in the repo

**Comment by janpio** [comment]
So it might not actually be a regression? Or does it work in 5.5.0?
If I understand your message correct, the repro repo does not actually reproduce, correct?

**Comment by cipriancaba** [comment]
It doesn't reproduce because the prisma client folder gets generated in the root of the repo `node_modules` and not in the local package (`packages/db/node_modules`). I am confused as to why that happens and was hoping you can help me understand

**Comment by SevInf** [comment]
@cipriancaba for starters, we'd really like to actually get a reproduction of the issue. Unfortinately, repo where issue does not happen does not help us either. Could you try to reduce you private repo to a minimum size where issue still reproduces but remove all actually private stuff from it and get something you can share?

Btw. one think worth checking is `. npmrc` file - does it have the same contents in your public repo and in your private one?

**Comment by Jolg42** [comment]
@cipriancaba Do you have time to take a look at @SevInf request? Thanks.

**Comment by senghuotlay** [comment]
what got me working is by doing this

```
package.json
{
  "name": "@package/database",
  "version": "1.0.0",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "tsup && npx prisma generate"
  },
  ```
  
  ```
  schema.prisma
  generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["fullTextSearch", "fullTextIndex"]
  output = "../dist"
}
```

**Comment by SevInf** [comment]
Hey @cipriancaba ,
 

Without a proper reproduction, we unfortunately have to close the issue for now. Feel free to comment if you manage to create one and we'll reopen the issue. 
 
Thank you 🙏

**Comment by wmtrinu** [comment]
Getting the same error with tsup and turborepo


```
PrismaClientInitializationError: Prisma Client could not locate the Query Engine for runtime "darwin-arm64".

This is likely caused by a bundler that has not copied "libquery_engine-darwin-arm64.dylib.node" next to the resulting bundle.
Ensure that "libquery_engine-darwin-arm64.dylib.node" has been copied next to the bundle or in "packages/api/dist".

We would appreciate if you could take the time to share some information with us.
Please help us by answering a few questions: https://pris.ly/engine-not-found-bundler-investigation
```

**Comment by cipriancaba** [comment]
Hi @Jolg42 . Monorepos are normally used in complex environments which are not easy to replicate. I've tried doing this but can't replicate it on a new project because in my actual setup the prisma client gets generated in:

```
packages/db/node_modules/.prisma
```

and while in my repro above it gets correctly generated in

```
node_modules/.prisma
```

Now if I were to understand how `prisma generate` chooses the location to generate the client by default, it might help with the repro case. 

While i know it would be really helpful for you to have a repro case, i feel this is a help me to help you situation as opposed to closing the issue. 

For anyone wondering, I am working around the issue with having a custom output location and using the stupid nextjs prisma plugin which is somethings causing infinite loops while prisma is trying to locate the query engine and just hangs forever

```
generator client {
  provider        = "prisma-client-js"
  output          = "../client"
  previewFeatures = ["multiSchema", "views", "relationJoins"]
  binaryTargets   = ["native", "darwin-arm64", "rhel-openssl-3.0.x"]
  engineType      = "library"
}
```

next.config.js
```
webpack: (config, { isServer }) => {
    if (isServer) {
      config.plugins = [...config.plugins, new PrismaPlugin()]
    }

    return config
  },
```

**Comment by cipriancaba** [comment]
@Jolg42 is there maybe a different way of importing the `libquery_engine-darwin-arm64.dylib.node` query engine? Maybe setting an ENV path or something that would reference it?

**Comment by Jolg42** [comment]
@cipriancaba Indeed there is that one `PRISMA_QUERY_ENGINE_LIBRARY=custom/libquery_engine-darwin-arm64.dylib.node`
Sounds interesting to try it out, good idea 😉 
https://www.prisma.io/docs/orm/reference/environment-variables-reference#prisma_query_engine_library

**Comment by idebeijer** [comment]
I ran into the same issue using the same monorepo provided above, and with the env variable @Jolg42 mentioned I got it to work. It feels really hacky but this did it for me:
```sh
PRISMA_QUERY_ENGINE_LIBRARY="../../node_modules/.pnpm/@prisma+client@5.15.0_prisma@5.15.0/node_modules/.prisma/client/libquery_engine-darwin-arm64.dylib.node"
```

**Comment by Jolg42** [comment]
Thanks @idebeijer for the update for the workaround. I think our problem here is that we would need to get a reproduction first before we can attempt a fix.

**Comment by sfratini** [comment]
@Jolg42 
I think I found one reason why this could happen. We have a mono repo and we are using esbuild. Since we need to keep some dependencies in the node_modules (pg, events), we were having trouble with the final dist file. So we ended up doing something lazy but simple. We just have another package.json with these dependencies so we can copy whatever node_modules folder is generated into a zip file for distribution. The structure is as follows (simplified):

- build
- - node_modules
- - package.json (this has pg, pg-native and prisma)
- node_modules (folder for local dev)
- src
- - prisma
- - - schema.prisma
- - package.json (this one has prisma again, and all of our other dependencies which will later be bundled by esbuild, but ignoring the build folder)

Now, in the beginning, we added the schema in the first package.json as "../src/prisma/schema.prisma" and "ran npx prisma generate" inside the "build" folder. The client was being generated but in the top parent node_modules. The one in build/node_modules/.prisma was not initialised and also was missing the engines. I was expecting prisma generate to run from the "build/node_modules" but it was not. 

In order to fix it, we just copy the schema.prisma into the build folder during the packing and then the client generates inside the right build/node_modules and we can copy the client to the final dist.zip file. I guess I was expecting the prisma generate to generate the client relative to where the script is ran but I believe it is doing it from the schema.prisma and the nearest node_modules

**Comment by sfratini** [comment]
Another possible problem, since we also had this is that we were using pm2 as a package manager. When running, pm2 changes the cwd so your app looks for the prisma client in a different path. More info here: https://github.com/Unitech/pm2/issues/96
So if you are using pm2 you need to set up your cwd as well (this is related to pm2, not Prisma) but we had that issue as well.

**Comment by felixbridell** [comment]
Ran in to this issue today when trying to setup a new monorepo using turborepo with the prisma example like:
`yarn dlx create-turbo@latest --example with-prisma`.

Did a search for the engine in the repo with `find ./ -name "libquery_engine-darwin-arm64.dylib.node" -type f -print0` and find it at these places:

```
./node_modules/prisma/engines/12e25d8d06f6ea5a0252864dd9a03b1bb51f3022/libquery_engine-darwin-arm64.dylib.node
./node_modules/prisma/libquery_engine-darwin-arm64.dylib.node
./node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node
./packages/database/node_modules/prisma/libquery_engine-darwin-arm64.dylib.node
./packages/database/node_modules/.prisma/client/libquery_engine-darwin-arm64.dylib.node
./packages/database/node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node
```
Did try to use the `PRISMA_QUERY_ENGINE_LIBRARY` to set the location manually as suggested above but that did not solve the problem.
